modules/lang-painless/src/yamlRestTest/resources/rest-api-spec/test/painless/165_dense_vector_byte_special_cases.yml

```diff
@@ -0,0 +1,245 @@
+setup:
+  - skip:
+      features: headers
+
+  - do:
+      indices.create:
+        index: test-index
+        body:
+          settings:
+            number_of_replicas: 0
+            number_of_shards: 1
+          mappings:
+            properties:
+              vector:
+                type: dense_vector
+                element_type: byte
+                dims: 3
+              indexed_vector:
+                type: dense_vector
+                element_type: byte
+                dims: 3
+                index: true
+                similarity: l2_norm
+                index_options:
+                  type: hnsw
+                  m: 15
+                  ef_construction: 50
+
+---
+"Indexing of Dense vectors should error when dims don't match defined in the mapping":
+
+  - do:
+      catch: bad_request
+      index:
+        index: test-index
+        id: "1"
+        body:
+          vector: [10, 2]
+  - match: { error.type: "mapper_parsing_exception" }
+
+  - do:
+      catch: bad_request
+      index:
+        index: test-index
+        id: "1"
+        body:
+          indexed_vector: [10, 2]
+  - match: { error.type: "mapper_parsing_exception" }
+
+---
+"Vectors of mixed integers and floats":
+  - do:
+      index:
+        index: test-index
+        id: "1"
+        body:
+          vector: [10, 10, 10]
+
+  - do:
+      index:
+        index: test-index
+        id: "2"
+        body:
+          vector: [11.0, 10.0, 10.0]
+
+  - do:
+      indices.refresh: {}
+
+  # query vector of type integer
+  - do:
+      headers:
+        Content-Type: application/json
+      search:
+        rest_total_hits_as_int: true
+        index: test-index
+        body:
+          query:
+            script_score:
+              query: {match_all: {} }
+              script:
+                source: "cosineSimilarity(params.query_vector, 'vector')"
+                params:
+                  query_vector: [10, 10, 10]
+
+  - match: {hits.total: 2}
+  - match: {hits.hits.0._id: "1"}
+  - match: {hits.hits.1._id: "2"}
+
+  # query vector of type float
```

Comment:
Good question!

Okay, so I went and added unit tests in `VectorScoreScriptUtilsTests` to cover the class-bindings portion of this, and invalid byte vectors are screened by `checkVectorBounds`, so there can be no invalid vectors here. 

However, for non-class bindings this gets a bit weird to support arity-only type overloading in Painless. (Note that if a user is doing this they are already doing something more advanced and less efficient, so we recommend they use the class-bindings which can cache the query vector and normalize when it makes sense to do so.)

The following can be passed into one of the vector calc methods on a byte vector:
byte[] - known to be valid
float[] - will throw an UnsupportedOperationException now
List<Number> - no checks and will convert anything to a byte

I think for the List<Number> case it would be *very* expensive to add bounds checking here. The user is already in a advanced use case, so I would advocate against this check in this singular case.